### ISSUE TITLE
Wrong Error Raised: "The graph couldn't be sorted in topological order"

### ISSUE THREAD

**Issue by haoransh** [tensorflow/tensorflow#24816]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip from anaconda
- TensorFlow version (use command below): 1.12.0
- Python version: 3.5.0
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 9.0
- GPU model and memory: GeForce GTX 1080Ti / Tesla K80

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**

The tensorflow will raise `The graph couldn't be sorted in topological order` Error when executing the optimizer. While The error doesn't occur on tensorflow 1.10.0. This error is also posed [here](https://stackoverflow.com/questions/54088172/why-tensorflow-throws-the-graph-couldnt-be-sorted-in-topological-order/54120941#54120941) by another user.

**Describe the expected behavior**

The error should not be raised because there is no loop in the computation graph.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

Reference: https://stackoverflow.com/questions/54088172/why-tensorflow-throws-the-graph-couldnt-be-sorted-in-topological-order/54120941#54120941
```
import tensorflow as tf
print(tf.__version__)
activation = tf.nn.relu
img_plh = tf.placeholder(tf.float32, [None, 3, 3, 3])
label_plh = tf.placeholder(tf.float32, [None])
layer = img_plh
buffer = []
ks_list = list(range(1, 10, 1)) + list(range(9, 0, -1))
for ks in ks_list:
    buffer.append(tf.layers.conv2d(layer, 9, ks, 1, "same", activation=activation))
layer = tf.concat(buffer, 3)
layer = tf.layers.conv2d(layer, 1, 3, 1, "valid", activation=activation)
layer = tf.squeeze(layer, [1, 2, 3])
loss_op = tf.reduce_mean(tf.abs(label_plh - layer))
optimizer = tf.train.AdamOptimizer()
train_op = optimizer.minimize(loss_op)
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    result = sess.run(train_op, {img_plh: np.zeros([2, 3, 3, 3], np.float32), label_plh: np.zeros([2], np.float32)})
```
**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by Gonzalo933** [comment]
I am experiencing this, also in tensorflow 1.12.0. I'm using a custom deep network and the same architecture sometimes raises this error and sometimes not (which makes me believe that is a tensorflow issue)

**Comment by qwesdfok** [comment]
I upgraded to 1.13.1 (from pip in python 3.6 & Windows 10). This problem still exists.
BUT it can be solved by move Activation Function to the place after tf.concat(buffer, 3)

`
for ks in ks_list:
buffer.append(tf.layers.conv2d(layer, 9, ks, 1, "same"))
layer = tf.concat(buffer, 3)
layer = activation(layer)
`

**Comment by taylorchu** [comment]
Could you try to decrease batch size? This seems to be memory related.

**Comment by qwesdfok** [comment]
> Could you try to decrease batch size? This seems to be memory related.

My GPU has 8GB memory, and the problem still exists. In the example, the batch size of the input is only 2. If it is a memory problem, the batch size does not seem to be the critical condition that causes the problem. In my response, the problem may be related to the implementation of the activator.

**Comment by Gonzalo933** [comment]
> > Could you try to decrease batch size? This seems to be memory related.
> 
> My GPU has 8GB memory, and the problem still exists. In the example, the batch size of the input is only 2. If it is a memory problem, the batch size does not seem to be the critical condition that causes the problem. In my response, the problem may be related to the implementation of the activator and the optimizer, because the error disappears after moving the activator out of the For loop.

I do think it has to be with memory and the graph even if not using all the RAM. My code is not a neural network per se and I'm not using any activation function, but when the number of unit in a layer is low I don't get the error, but running the **same code** with more units in the hidden layers causes the error to appear, so I think that the error may be cause by the graph growing too much to do any optimization that TF normally does. But that is just my guess

**Comment by qwesdfok** [comment]
> > > Could you try to decrease batch size? This seems to be memory related.
> > 
> > 
> > My GPU has 8GB memory, and the problem still exists. In the example, the batch size of the input is only 2. If it is a memory problem, the batch size does not seem to be the critical condition that causes the problem. In my response, the problem may be related to the implementation of the activator and the optimizer, because the error disappears after moving the activator out of the For loop.
> 
> I do think it has to be with memory and the graph, even if not using all the RAM. My code is not a neural network per se and I'm not using any activation function, but when the number of units in a layer is low I don't get the error. Running the **same code** with more units in the hidden layers causes the error to appear. I think that the error may be caused by the graph growing too much to do any optimization that TF normally does. But that is just my guess

I think your guess is right.

**Comment by taylorchu** [comment]
> I do think it has to do with memory and the graph, even if not using all the RAM. My code is not a neural network per se and I'm not using any activation function, but when the number of units in a layer is low I don't get the error. Running the **same code** with more units in the hidden layers causes the error to appear. I think that the error may be caused by the graph growing too much to do any optimization that TF normally does. But that is just my guess

For me, a *12-layer* network does not have this error, but  a *20-layer* network will have this error.
In addition, if you see this error, the weights updated are incorrect.

**Comment by zhanghaojia123456** [comment]
I met the same problem. When double-check the updated weights there is slightly error after updating.

**Comment by fonspa** [comment]
Any update on this issue ? I got the same problem here with Tensorflow 2.0.0alpha0 and beta1-gpu or cpu on Ubuntu 18.04 LTS and Windows 10.
Tried to reduce batch size, to move activations out of loops or to reduce number of layers / units to no avail.
My model / dataset works _fine_ with TF 1.9.0-gpu without this warning, and the time per step is even lower by more than 33% when compared to an execution on TF 2.0 (which gives the warning).
Thanks.

**Comment by panfengde** [comment]
I want to build a SSD netWork by myself ,and I meet the same problem .
![image](https://user-images.githubusercontent.com/27055350/60565650-04f07200-9d97-11e9-8d35-2db98f9ce4ab.png)


i have seen this method,it can run ok.
![image](https://user-images.githubusercontent.com/27055350/60565680-20f41380-9d97-11e9-944f-1daeb9d330dd.png)
then i try to solve my problem like this method!
the old code as show belown
![image](https://user-images.githubusercontent.com/27055350/60565766-64e71880-9d97-11e9-8435-3bc5adac7a29.png)
i removed the Activation function，like this:
![image](https://user-images.githubusercontent.com/27055350/60565914-e474e780-9d97-11e9-9eda-4f28867e7677.png) 
the proplem disappeared
conclusion：

whten many tensors which are actived  by Activation function are  concated  as a list ,this problem will appear!
i think this is  a problem of tensorflow.

**Comment by beihaifusang** [comment]
confirmed again that it is the problem of tensorflow.

In implementing spp layer by slicing and when used with dense bins (e.g. a 27*27 pooled feature map), tensorflow complains when the 27*27=729 tensors are concatenated in one concat op.

But it is confusing that it complains about the wrong topological order and such error message is misleading. 

If such operation exceeds the current capability of graph optimizer, could it generate a message like "operations with too much tensor", instead of complaining on the graph topology?

**Comment by fonspa** [comment]
Hm, while initially I thought my problem was related to this issue I know now it has nothing to do with concatenate layers but with my custom loss function. I will open a new issue. Thanks.

**Comment by Pay20Y** [comment]
For me it appear when a manual loop with too many steps  is used in my graph.

**Comment by pachirayz** [comment]
Any news&process here? I also face the same problem.
Here is my code:

`weight_atom = tf.get_variable(name, [2, 2])`
`#do repetition`
    `weight_row = tf.concat([weight_atom] * 2, axis=1)`
    `weight = tf.concat([weight_row] * 2, axis=0)`
   ` out = tf.matmul(input, weights) `

The code create a (2,2) matrix then does repetition and multiply the result with the input which shape is (4,4)

**Comment by OliPhilip** [comment]
Same error occurs **only for validation** using identical generators for train and valid for the dummy example below (tf 1.14, python 3.6):

[tf_topo_bug.py.txt](https://github.com/tensorflow/tensorflow/files/3613997/tf_topo_bug.py.txt)

**Comment by liyi193328** [comment]
Any update? Face the same problem too.

**Comment by huangxGo** [comment]
> confirmed again that it is the problem of tensorflow and I'm using r1.13.
> 
> In implementing spp layer by slicing and when used with dense bins (e.g. a 27x27 pooled feature map), tensorflow complains when the 27x27=729 tensors are concatenated in one concat op.
> 
> But it is confusing that it complains about the wrong topological order and such error message is misleading.
> 
> If such operation exceeds the current capability of graph optimizer, could it generates a message like "operations with too much tensor", instead of complaining about the graph topology?

I have the same problem when I use spp layer with tensorflow 1.14.If it is right when I  use spp layer and just ignore this misleading message?

**Comment by Gonzalo933** [comment]
> > confirmed again that it is the problem of tensorflow and I'm using r1.13.
> > In implementing spp layer by slicing and when used with dense bins (e.g. a 27x27 pooled feature map), tensorflow complains when the 27x27=729 tensors are concatenated in one concat op.
> > But it is confusing that it complains about the wrong topological order and such error message is misleading.
> > If such operation exceeds the current capability of graph optimizer, could it generates a message like "operations with too much tensor", instead of complaining about the graph topology?
> 
> I have the same problem when I use spp layer with tensorflow 1.14.If it is right when I use spp layer and just ignore this misleading message?

I have detected that after this warning performance is worse

**Comment by huangxGo** [comment]
> > > confirmed again that it is the problem of tensorflow and I'm using r1.13.
> > > In implementing spp layer by slicing and when used with dense bins (e.g. a 27x27 pooled feature map), tensorflow complains when the 27x27=729 tensors are concatenated in one concat op.
> > > But it is confusing that it complains about the wrong topological order and such error message is misleading.
> > > If such operation exceeds the current capability of graph optimizer, could it generates a message like "operations with too much tensor", instead of complaining about the graph topology?
> > 
> > 
> > I have the same problem when I use spp layer with tensorflow 1.14.If it is right when I use spp layer and just ignore this misleading message?
> 
> I have detected that after this warning performance is worse

Yes it seems that we can't ignore this warning...In my training, the loss get very high after a few iterations

**Comment by beihaifusang** [comment]
> > confirmed again that it is the problem of tensorflow and I'm using r1.13.
> > In implementing spp layer by slicing and when used with dense bins (e.g. a 27x27 pooled feature map), tensorflow complains when the 27x27=729 tensors are concatenated in one concat op.
> > But it is confusing that it complains about the wrong topological order and such error message is misleading.
> > If such operation exceeds the current capability of graph optimizer, could it generates a message like "operations with too much tensor", instead of complaining about the graph topology?
> 
> I have the same problem when I use spp layer with tensorflow 1.14.If it is right when I use spp layer and just ignore this misleading message?

Better not. Instead, you could try to do multiple concat instead of a huge concat. (yes, that's ugly)

**Comment by sh1ng** [comment]
Faced it when use `concat` as `repeat_elements`

**Comment by LukeALee** [comment]
I faced the same problem when I use
`self.embeddingSentences = tf.reshape(tf.concat(self.embeddingSentences, axis=0), [config.batchSize, -1, config.model.hiddenSizes[-1] * 2])`  in Ubuntu 18.04, 12G GPU RAM,
to concat a list length 64  of 2D tensors which size were [15, 512]  to reshape it to a 3D tensors of [64, 15, 512] to be the batch input of the following Bi-LSTM network.

After see the comment of 

> I faced the same issue because of a huge concat. Thanks to the comments above, I could work around it and avoid calling concat which resolved the issue for me.

I change the code to 
`self.embeddingSentences = tf.reshape(self.embeddingSentences, [config.batchSize, -1, config.model.hiddenSizes[-1] * 2])`
even though if it was the right useage, **but this error disappeared.**

so I do think this error will be raised when we use huge concat.

**Comment by liushengzhong1023** [comment]
I got the same issue, then I tried to replace the `tf.concat` with `tf.stack` function and the issue disappeared. I am doing so because I am concatenating into a new dimension. Hope this can help.

**Comment by colinwke** [comment]
I implement a custom `tf.feature_column.input_layer` and encounter the problem.
and I reference the official implement, and add a `tf.reshape` to each tensor before concat, then the error disappeared!

**Comment by beihaifusang** [comment]
> I implement a custom `tf.feature_column.input_layer` and encounter the problem.
> and I reference the official implement, and add a `tf.reshape` to each tensor before concat, then the error disappeared!

Hi Colin, that's interesting, could you share more about your findings about this?

**Comment by colinwke** [comment]
> > I implement a custom `tf.feature_column.input_layer` and encounter the problem.
> > and I reference the official implement, and add a `tf.reshape` to each tensor before concat, then the error disappeared!
> 
> Hi Colin, that's interesting, could you share more about your findings about this?

Hi @beihaifusang 
this is part of my code, and bottom comment contain more info, hope to be helpful to you.

```python


def build_input_layer(feature_spec, features):
    with tf.variable_scope(name_or_scope="input_layer"):
        output_tensors = []
        for name, prop in feature_spec.items():
            if "num_unique" in prop["attr"] and prop["attr"]["num_unique"] <= 500:
                tensor = tf.one_hot(features[name], depth=prop["attr"]["num_unique"], name="one_hot_" + name)
                batch_size = array_ops.shape(tensor)[0]
                tensor = array_ops.reshape(
                    tensor, shape=(batch_size, prop["attr"]["num_unique"]))
                output_tensors.append(tensor)
            else:
                bucket_size = 10 ** 3
                embedding_size = 32
                num_unique = prop["attr"]["num_unique"] if "num_unique" in prop["attr"] else 0

                embeddings = tf.get_variable(name="emb" + name,
                                             dtype=tf.float32,
                                             shape=[1000, 32])
                tensor = tf.nn.embedding_lookup(
                    embeddings, features[name],
                    name="lookup_table_" + name)  # shape(batch_size, max_seq_len, embedding_size)
                batch_size = array_ops.shape(tensor)[0]
                tensor = array_ops.reshape(
                    tensor, shape=(batch_size, 32))
                output_tensors.append(tensor)

        """
        log error:
        E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:666] Iteration = 0, topological sort failed with message: The graph couldn't be sorted in topological order.
        reference code:
        locate: tensorflow/python/feature_column/feature_column.py

            ...
            batch_size = array_ops.shape(tensor)[0]
            output_tensor = array_ops.reshape(
                tensor, shape=(batch_size, num_elements))
            output_tensors.append(output_tensor)
            if cols_to_vars is not None:
              # Retrieve any variables created (some _DenseColumn's don't create
              # variables, in which case an empty list is returned).
              cols_to_vars[column] = ops.get_collection(
                  ops.GraphKeys.GLOBAL_VARIABLES,
                  scope=variable_scope.get_variable_scope().name)
            if cols_to_output_tensors is not None:
              cols_to_output_tensors[column] = output_tensor
        _verify_static_batch_size_equality(output_tensors, ordered_columns)
        return array_ops.concat(output_tensors, 1)
        """

    return array_ops.concat(output_tensors, 1, name="concat_input_layer")
```

**Comment by andreped** [comment]
I got this error as I was using tf.keras.backend.repeat_elements (TF=1.13.1). If the tensor is of type int32/int64, it is possible to use tf.tile instead. 

In my case, I wanted to repeat and concat a binary 1-dim tensor, A, to be used for filtering another tensor. I managed to get the same result as with repeat_elements without the error with:
`A = tf.expand_dims(A, axis=1)`
'A = tf.tile(A, [1, num_repeats])'

It should also be possible to run tf.tile and then tf.reshape, but at least in my simple case what I wrote above was sufficient. tf.tile also supports higher dims:
https://www.tensorflow.org/api_docs/python/tf/tile

**Comment by Osos8128** [comment]
Any updates how to solve this problem ??

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.

**Comment by Mastiff37** [comment]
For what it's worth, I have a similar issue in TF 2.0.0 with concatenate and K.repeat_elements.   Nothing complicated and it should work, but complained about sorting and gave other errors when attempting to predict or train.  In my case I was able to eliminate the repeat elements and allow automatic broadcasting to do the equivalent thing, but this won't always be possible.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, Please open a new issue for any help you need against 2.x, and we will get you the right help.
